src/server/stream_family.cc

```diff
@@ -2543,46 +2588,52 @@ bool ParseXpendingOptions(CmdArgList& args, PendingOpts& opts, SinkReplyBuilder*
 }  // namespace
 
 void StreamFamily::XAdd(CmdArgList args, const CommandContext& cmd_cntx) {
-  auto parse_resp = ParseAddOrTrimArgsOrReply(args, true, cmd_cntx.rb);
-  if (!parse_resp) {
+  CmdArgParser parser{args};
+  auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
+
+  string_view key = parser.Next();
+
+  auto parsed_add_opts = ParseAddOptsOrReply(&parser, rb);
+  if (!parsed_add_opts) {
     return;
   }
 
-  auto add_opts = parse_resp->first;
-  auto id_indx = parse_resp->second;
-
-  args.remove_prefix(id_indx);
-  if (args.size() < 2 || args.size() % 2 == 0) {
-    return cmd_cntx.rb->SendError(WrongNumArgsError("XADD"), kSyntaxErrType);
+  if (auto err = parser.Error(); err) {
```

Comment:
this is a bit confusing why in some cases ParseAddOptsOrReply sends the error reply and some cases it is sent here.
I suggest make ParseAddOptsOrReply just parser (will not get the reply builder) it will return the parsed args and in case of an error return the error and the error will be sent only here